stdenv: allow preserving meta fields in derivation

Princemachiavelli opened the pull request:
## Changes
* Preserve `identifiers`, `license`, `mainProgram`, and `vendor` fields from the `mkDerivation` attrSet under a new `nixMeta` field in the derivation.
* `nixMeta` respects `__structuredAttrs`; if enabled the existing attribute structure of the aforementioned attributes is preserved. If `__structuredAttrs` is false, the attributes will be encoded as a JSON string.
* A evaluation error will occur if any string-context is present in `nixMeta`. 

<details>
<summary> Example Derivation output </summary>

For example, a derivation's `env` will look like this if `__structuredAttrs` is false:
```json
{
  "/nix/store/aflkbqqs7f8hyp0l6przb3bpxr5xs1dh-hello-2.12.2.drv": {
    "args": [
      "-e",
      "/nix/store/l622p70vy8k5sh7y5wizi5f2mic6ynpg-source-stdenv.sh",
      "/nix/store/shkw4qm9qcw5sc5n1k5jznc83ny02r39-default-builder.sh"
    ],
    "builder": "/nix/store/mfy6cpswaknpqj1jv7h5s27dh60ziv3p-bash-5.3p3/bin/bash",
    "env": {
      "nixMeta": "{\"identifiers\":{\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"vendor\":\"gnu\"},\"possibleCPEs\":[{\"cpe\":\"cpe:2.3:a:gnu:hello:2.12.2:*:*:*:*:*:*:*\",\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"update\":\"*\",\"vendor\":\"gnu\",\"version\":\"2.12.2\"}},{\"cpe\":\"cpe:2.3:a:gnu:hello:2.12:2:*:*:*:*:*:*\",\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"update\":\"2\",\"vendor\":\"gnu\",\"version\":\"2.12\"}}],\"v1\":{\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"vendor\":\"gnu\"},\"possibleCPEs\":[{\"cpe\":\"cpe:2.3:a:gnu:hello:2.12.2:*:*:*:*:*:*:*\",\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"update\":\"*\",\"vendor\":\"gnu\",\"version\":\"2.12.2\"}},{\"cpe\":\"cpe:2.3:a:gnu:hello:2.12:2:*:*:*:*:*:*\",\"cpeParts\":{\"edition\":\"*\",\"language\":\"*\",\"other\":\"*\",\"part\":\"a\",\"product\":\"hello\",\"sw_edition\":\"*\",\"target_hw\":\"*\",\"target_sw\":\"*\",\"update\":\"2\",\"vendor\":\"gnu\",\"version\":\"2.12\"}}]}},\"license\":{\"deprecated\":false,\"free\":true,\"fullName\":\"GNU General Public License v3.0 or later\",\"redistributable\":true,\"shortName\":\"gpl3Plus\",\"spdxId\":\"GPL-3.0-or-later\",\"url\":\"https://spdx.org/licenses/GPL-3.0-or-later.html\"},\"mainProgram\":\"hello\"}",
      "out": "/nix/store/23p192rg0c8r0lzqgrr0gykk6vlwlzj5-hello-2.12.2",
      [...]
      "system": "x86_64-linux",
      "version": "2.12.2"
    },
    "inputDrvs": {},
    [...]
}
```
If `__structuredAttrs` is enabled, then `nixMeta` is added as a attrSet.
```json
{
  "/nix/store/5gpn6haap01zy2l8xdnx84my6lafshkn-hello-2.12.2.drv": {
    "args": [
      "-e",
      "/nix/store/l622p70vy8k5sh7y5wizi5f2mic6ynpg-source-stdenv.sh",
      "/nix/store/shkw4qm9qcw5sc5n1k5jznc83ny02r39-default-builder.sh"
    ],
    "builder": "/nix/store/mfy6cpswaknpqj1jv7h5s27dh60ziv3p-bash-5.3p3/bin/bash",
    "env": {
      "out": "/nix/store/g45fpaampg8hfwf8cj61y8pl1nbjwsfs-hello-2.12.2"
    },
    "inputDrvs": {
     [...]
      }
    },
    "inputSrcs": [...],
    "name": "hello-2.12.2",
    "outputs": {
      "out": {
        "path": "/nix/store/g45fpaampg8hfwf8cj61y8pl1nbjwsfs-hello-2.12.2"
      }
    },
    "structuredAttrs": {
      [...]
      "name": "hello-2.12.2",
      "nativeBuildInputs": [
        "/nix/store/671cssvzsgf47jqxm8zcl7lgpnbqivm3-version-check-hook"
      ],
      "nixMeta": {
        "identifiers": {
          "cpeParts": {
            "edition": "*",
            "language": "*",
            "other": "*",
            "part": "a",
            "product": "hello",
            "sw_edition": "*",
            "target_hw": "*",
            "target_sw": "*",
            "vendor": "gnu"
          },
          "possibleCPEs": [
            {
              "cpe": "cpe:2.3:a:gnu:hello:2.12.2:*:*:*:*:*:*:*",
              "cpeParts": {
                "edition": "*",
                "language": "*",
                "other": "*",
                "part": "a",
                "product": "hello",
                "sw_edition": "*",
                "target_hw": "*",
                "target_sw": "*",
                "update": "*",
                "vendor": "gnu",
                "version": "2.12.2"
              }
            },
            {
              "cpe": "cpe:2.3:a:gnu:hello:2.12:2:*:*:*:*:*:*",
              "cpeParts": {
                "edition": "*",
                "language": "*",
                "other": "*",
                "part": "a",
                "product": "hello",
                "sw_edition": "*",
                "target_hw": "*",
                "target_sw": "*",
                "update": "2",
                "vendor": "gnu",
                "version": "2.12"
              }
            }
          ],
          "v1": {
            "cpeParts": {
              "edition": "*",
              "language": "*",
              "other": "*",
              "part": "a",
              "product": "hello",
              "sw_edition": "*",
              "target_hw": "*",
              "target_sw": "*",
              "vendor": "gnu"
            },
            "possibleCPEs": [
              {
                "cpe": "cpe:2.3:a:gnu:hello:2.12.2:*:*:*:*:*:*:*",
                "cpeParts": {
                  "edition": "*",
                  "language": "*",
                  "other": "*",
                  "part": "a",
                  "product": "hello",
                  "sw_edition": "*",
                  "target_hw": "*",
                  "target_sw": "*",
                  "update": "*",
                  "vendor": "gnu",
                  "version": "2.12.2"
                }
              },
              {
                "cpe": "cpe:2.3:a:gnu:hello:2.12:2:*:*:*:*:*:*",
                "cpeParts": {
                  "edition": "*",
                  "language": "*",
                  "other": "*",
                  "part": "a",
                  "product": "hello",
                  "sw_edition": "*",
                  "target_hw": "*",
                  "target_sw": "*",
                  "update": "2",
                  "vendor": "gnu",
                  "version": "2.12"
                }
              }
            ]
          }
        },
        "license": {
          "deprecated": false,
          "free": true,
          "fullName": "GNU General Public License v3.0 or later",
          "redistributable": true,
          "shortName": "gpl3Plus",
          "spdxId": "GPL-3.0-or-later",
          "url": "https://spdx.org/licenses/GPL-3.0-or-later.html"
        },
        "mainProgram": "hello"
      },
      "outputChecks": {
        "out": {}
      },
      "outputs": [
        "out"
      ],
      "patches": [],
      "pname": "hello",
   [...]
  }
```
</details>

<details>
<summary> Proposal Background </summary>

Currently, package metadata is stripped before creating the derivation. Generally, we only retain the package name and version in the derivation and we loosely retain this is the nix store paths except reconstructing the exact name and version is imprecise.

This lost of meta information makes generating correct and complete SBOMs nearly impossible. Tools like `sbomnix` require matching store paths and derivations to the correct `meta` information solely on the package name and version which often leads to incorrect matches and incomplete SBOM reports.

An existing PR (#409797)  to improve the state of CVE tracking will add the complete CPE identifier to the `meta` attribute set.  While this is a great improvement, nixpkgs still lacks method to query the meta/CPE information for a given derivation or store-path. A previous PR #256296, attempted to resolve this by including the *entire* `meta` attrSet in the derivation but concerns over the risk of mass-rebuilds due to minor changes (such as to the maintainers list) stalled it's adoption.

Instead of including all meta attributes, this PR includes subset of meta attributes; `identifiers`, `license`, `mainProgram`, and `vendor` . Changes to these attributes are in-frequent or would involve a rebuild anyway (e.g updating `mainProgram`).
</details>

Princemachiavelli committed:
Allow preserving meta fields in derivation

Aleksanaa commented:
So we'll have a list of meta attribute which gets added to the derivation as well. And `mainProgram`.

It would be nice if we can come up with another name (other than `meta`) to place them, I think.

ConnorBaker commented on pkgs/build-support/setup-hooks/write-meta-json.sh:
```diff
@@ -0,0 +1,38 @@
+# This setup hook, writes the derivations nixMetaJSON info
+# to $output/nix-support/meta.json so package information
```
```suggestion
# This setup hook writes the derivation's `nixMetaJSON`
# to $output/nix-support/meta.json, so package information
```

ConnorBaker commented on pkgs/build-support/setup-hooks/write-meta-json.sh:
```diff
@@ -0,0 +1,38 @@
+# to $output/nix-support/meta.json so package information
+# can be reconstructed at runtime.
+
+
```
```suggestion
```

ConnorBaker commented on pkgs/build-support/setup-hooks/write-meta-json.sh:
```diff
@@ -0,0 +1,38 @@
+  return 0
+fi
+declare -ig writeMetaJSONInstalled=1
+
```
```suggestion
```

ConnorBaker commented:
@Princemachiavelli I'd like to see a check or assertion on the values being serialized into the derivation which prevents addition of any paths, derivations, or strings with context -- I believe allowing any of those would cause their closures to be propagated forever.

fricklerhandwerk commented:
People likely interested in changes to `stdenv`: @infinisil @roberth @ericson2314

Princemachiavelli commented:
@fricklerhandwerk 
> My two cents: @Princemachiavelli do I understand correctly that you want this because `bombom` only reads store paths but not expressions? Because there are at least two more approaches that will achieve the original goal:

Yes, at a high level this is the issue. My main argument for this PR's solution is it's simplicity and minimal impact on Nix/Nixpkgs. I think both of your proposed approaches have a lot of interesting use-cases but, given their complexity, they should be implemented as part of a larger project to improve many parts of Nix/Nixpkgs.

> 1. Change Nix to put sidecar information into store objects that doesn't affect the store path

This is a proposed feature in upstream Nix (https://github.com/NixOS/nix/issues/10780). While this is an interesting solution, it presents some obvious issues. If CPE (or any metadata) is incorrect and must be updated, how would we "force" an update of a store path if we are excluding this data from the input-addressing function? The "actual" problem trying to be solved here is avoiding rebuilds due to metadata changes. However, one could argue that the same issue occurs due to minor changes to the `fixupPhase` or `checkPhase`. The fact that the expensive buildPhase is coupled with cheap post-processing steps is a much larger problem to carving out a new Nix feature just for this single use case seems like a lot of work for little gain. IMO I'd rather advocate for efficient per-phase derivations to allow for granular build caching.

> 2. Do everything at the expression level (i.e. "a package is files with metadata") and improve eval performance in Nix and type safety in Nixpkgs, as well as add tooling to make that nice.

I'm not 100% sure I'm following but I could imagine a solution that operates on the expression layer. Let me know if this tracks what you are thinking. I could see `mkDerivation` returning a second `meta` attribute-derivation for each orginial `output`. When this `meta` derivation is built, the `meta` information is preserved and the `buildInputs`, `nativeBuildInputs`, etc. are replaced with the `meta` equivelent of each. The actual buildPhase would just be building a tree structure of the all the `meta` fields. 

However, I feel that correctly accounting for stringContext and identifying run-time vs build-time dependencies would become very complicated.  Even correctly implemented, it would "require" access to the Nix source code and original expression - it would *never* allow for computing the SBOM of an arbitrary store path. The association of store path and Nix expression would have to be tracked outside Nix as well. This would severely limit integration into existing security, CVE scanning, etc. software.

roberth reviewed (changes requested):
Thank you for taking this initiative @Princemachiavelli!

I think it's unfortunate that we don't have a good opportunity to integrate `meta` cleverly into the derivation format without the rebuild trouble, which is feasible. We may get around to it one day, along with other changes.
Before such a future, I agree it's better to deliver *a* solution that works, and this approach supports that.

arianvp commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -714,6 +716,15 @@ let
+        "license"
+        "vendor"
+        "cpe"
+        "mainProgram"
```
> In other words, tools should take into account the derivation closure, not just the output's closure. By storing the meta info in the derivation closure, we provide it faithfully in its original shape, and we set up our users to consider how to deal with these characteristics, instead of nudging them down the wrong path.

I don't understand the exclusive OR here.    This PR stores the info both in the derivation (all derivation arguments are exposed in the `.drv` after all) as well as in the output. It give you both.

I think having meta information of your *runtime* dependencies can  still be useful. e.g. for a runtime threat scanner to be able to tell you if you're running a vulnerable version of something. Or for a container scanner that scans a nix-built container to match store paths to package databases.  Then afterwards people can always reach for the derivation-level to get the whole accurate build-time view.

> Even using derivations, do all language ecosystems (rust, go, npm, etc.) have a single derivation per-dependency? If not, we would encounter the same issue.

`buildGoModule` and friends should be responsible for customizing the meta hook to output their SBOM information in `nix-support/`.    Go actually already stores dependency information of what it was built with inside the binary. So the meta information is already available.

```
% nix shell nixpkgs#go nixpkgs#caddy
% go version -m $(which caddy)
/nix/store/k369dl06sznpappxj4wr5fmh84x0zpjf-caddy-2.10.0/bin/caddy: go1.24.4
	path	github.com/caddyserver/caddy/v2/cmd/caddy
	mod	github.com/caddyserver/caddy/v2	(devel)	
	dep	cel.dev/expr	v0.19.1	
	dep	dario.cat/mergo	v1.0.1	
	dep	filippo.io/edwards25519	v1.1.0	
	dep	github.com/BurntSushi/toml	v1.4.0	
	dep	github.com/KimMachineGun/automemlimit	v0.7.1	
	dep	github.com/Masterminds/goutils	v1.1.1	
(...)
	build	-buildmode=exe
	build	-compiler=gc
	build	-tags=nobadger,nomysql,nopgx
	build	-trimpath=true
	build	CGO_ENABLED=1
	build	GOARCH=arm64
	build	GOOS=darwin
```

Bombon also has some code that does something similar for rust: https://github.com/nikstur/bombon/blame/main/nix/passthru-vendored.nix

The same we would have to do for statically linked binaries. Ideally the output of the statically built derivation contains meta information on what dependencies ended up being linked into the static binary (though modulo the nix store path. as we don't want references in static outputs I guess?). Similar to how we find propagated build inputs recursively for all our build inputs in `setup.sh`, we can also collect meta info recursively of all out build inputs and then propagate the union of that as a build output.  This might over-approximate as maybe not all `buildInputs` are used for linking in the end. But if we want exact info then this needs collaboration with the build tools of said packages to produce these info instead. See e.g. https://ariadne.space/2025/02/08/c-sboms-and-how-pkgconf.html

arianvp replied:
buildGoModule and friends could have their own setup hooks that add their own specific meta json with information we don't get at the derivation level and have those merged as well.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/list-all-installed-packages-version-for-audit-purposes/67280/5

Princemachiavelli committed:
bash: fix cpeParts

The bash version string may not be the expected format until
the fixed point is fully evaluated.

Princemachiavelli commented:
Performance impact (Nix 2.30.1):
----
| Derivation | Thunks Before PR | Thunks After PR | % Change |
|--|--|--|--|
| legacyPackages.x86_64-linux.linuxPackages_latest.kernel | 1259730 | 1293049 | 2.64 |
iso_graphical |  117783731 | 121095124 | 2.81 |
-----

Princemachiavelli committed:
Document addition of `nixMeta` field

Princemachiavelli committed:
mkDerivation: encode nixMeta as attrSet if __structuredAttrs is true

ConnorBaker commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -314,6 +314,9 @@
 
 - `meta.mainProgram`: Changing this `meta` entry can lead to a package rebuild due to being used to determine the `NIX_MAIN_PROGRAM` environment variable.
 
+-  mkDerivation populates a `nixMeta` field containing the following `meta` fields if present: `mainProgram`, `license`, `vendor`, `identifiers`.
```
```suggestion
-  mkDerivation populates a `nixMeta` field  on the derivation containing the following `meta` fields if present: `mainProgram`, `license`, `vendor`, `identifiers`.
```

Probably good to mention that `nixMeta` will exist as part of the derivation (and all that entails -- e.g., visible to the build environment, etc.).

emilazy commented:
I’d be really reluctant to extend the amount by which we mix `meta` and other derivation fields after the freeze, given the previous issues we’ve had with eval problems on https://github.com/NixOS/nixpkgs/pull/454333 and the concerns raised about the new behaviour for `meta.mainProgram` on https://github.com/NixOS/nixpkgs/pull/412768 (see also https://github.com/NixOS/nixpkgs/pull/430810). I think this should definitely wait for branch‐off even if consensus on the approach is reached.

Princemachiavelli committed:
Use intersectAttrs

Princemachiavelli committed:
Remove unnecessary string-context check

Princemachiavelli commented on pkgs/shells/bash/5.nix:
```diff
@@ -278,7 +278,7 @@ lib.warnIf (withDocs != null)
       identifiers.cpeParts =
         let
-          versionSplit = lib.split "p" version;
+          versionSplit = lib.split "p" fa.version;
```
This PR *does* require that `meta` successfully evaluate. I'm not sure why, but `version` here somehow has a strange value like "20250101-dirty". Pinning the version to the final fixed point value resolves the issue.
